File: packages/kotlin/src/test/kotlin/com/ecollect/sdk/CustomersModuleTest.kt
package com.ecollect.sdk

import com.ecollect.sdk.exceptions.*
import com.ecollect.sdk.modules.CustomersModule
import com.ecollect.sdk.modules.SessionModule
import com.ecollect.sdk.utils.HttpClient
import io.mockk.*
import kotlinx.coroutines.test.runTest
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test

class CustomersModuleTest {

    private lateinit var mockWebServer: MockWebServer
    private lateinit var httpClient: HttpClient

    @Before
    fun setUp() {
        mockWebServer = MockWebServer()
        mockWebServer.start()
        httpClient = HttpClient()
    }

    @After
    fun tearDown() {
        mockWebServer.shutdown()
    }

    private fun makeConfig(sessionToken: String = "test-session") = Config(
        etyCode = 123,
        sessionToken = sessionToken,
        environment = Environment.TEST,
        testBaseUrl = mockWebServer.url("/").toString()
    )

    private fun makeModule(sessionToken: String = "test-session"): CustomersModule {
        val config = makeConfig(sessionToken)
        val session = mockk<SessionModule>()
        coEvery { session.getSessionToken() } returns sessionToken
        every { session.invalidate() } just Runs
        return CustomersModule(config, httpClient, session)
    }

    private fun enqueueSuccess(customerId: String = "cust-001") {
        mockWebServer.enqueue(
            MockResponse()
                .setBody("""{"ReturnCode":"SUCCESS","CustomerInfoArray":[{"AttributeCode":37,"AttributeDesc":"CustomerId","AttributeValue":"$customerId"}]}""")
                .setResponseCode(200)
        )
    }

    @Test
    fun `getOrCreateCustomerId returns customerId for new customer`() = runTest {
        enqueueSuccess("cust-new-001")
        val module = makeModule()

        val result = module.getOrCreateCustomerId(
            usermail = "user@example.com",
            cardHolderId = "12345678",
            cardHolderName = "Juan Perez",
            cardHolderIdType = "CC",
            mobileCountryCode = "57",
            mobileNumber = "3001234567"
        )
        assertEquals("cust-new-001", result)
    }

    @Test
    fun `getOrCreateCustomerId returns same customerId for existing customer`() = runTest {
        enqueueSuccess("cust-existing-999")
        val module = makeModule()

        val result = module.getOrCreateCustomerId(
            usermail = "user@example.com",
            cardHolderId = "12345678",
            cardHolderName = "Juan Perez",
            cardHolderIdType = "CC",
            mobileCountryCode = "57",
            mobileNumber = "3001234567"
        )
        assertEquals("cust-existing-999", result)
    }

    @Test
    fun `updateCustomer returns customerId on success`() = runTest {
        enqueueSuccess("cust-upd-001")
        val module = makeModule()

        val result = module.updateCustomer(
            customerId = "cust-upd-001",
            usermail = "new@example.com"
        )
        assertEquals("cust-upd-001", result)
    }

    @Test
    fun `getOrCreateCustomerId throws CustomerNotFoundException when CustomerId missing in response`() = runTest {
        mockWebServer.enqueue(
            MockResponse()
                .setBody("""{"ReturnCode":"SUCCESS","CustomerInfoArray":[]}""")
                .setResponseCode(200)
        )
        val module = makeModule()

        try {
            module.getOrCreateCustomerId(
                usermail = "user@example.com",
                cardHolderId = "12345678",
                cardHolderName = "Juan Perez",
                cardHolderIdType = "CC",
                mobileCountryCode = "57",
                mobileNumber = "3001234567"
            )
            fail("Should throw CustomerNotFoundException")
        } catch (e: CustomerNotFoundException) {
            assertNotNull(e.message)
        }
    }

    @Test
    fun `getOrCreateCustomerId throws ValidationException for invalid card holder ID`() = runTest {
        mockWebServer.enqueue(
            MockResponse()
                .setBody("""{"ReturnCode":"FAIL_CARDHOLDERID"}""")
                .setResponseCode(200)
        )
        val module = makeModule()

        try {
            module.getOrCreateCustomerId(
                usermail = "user@example.com",
                cardHolderId = "bad-id",
                cardHolderName = "Juan Perez",
                cardHolderIdType = "CC",
                mobileCountryCode = "57",
                mobileNumber = "3001234567"
            )
            fail("Should throw ValidationException")
        } catch (e: ValidationException) {
            assertNotNull(e.message)
        }
    }

    @Test
    fun `getOrCreateCustomerId throws AuthenticationException on FAIL_ACCESSDENIED`() = runTest {
        mockWebServer.enqueue(
            MockResponse()
                .setBody("""{"ReturnCode":"FAIL_ACCESSDENIED"}""")
                .setResponseCode(200)
        )
        val module = makeModule()

        try {
            module.getOrCreateCustomerId(
                usermail = "user@example.com",
                cardHolderId = "12345678",
                cardHolderName = "Juan Perez",
                cardHolderIdType = "CC",
                mobileCountryCode = "57",
                mobileNumber = "3001234567"
            )
            fail("Should throw AuthenticationException")
        } catch (e: AuthenticationException) {
            assertNotNull(e.message)
        }
    }
}

File: packages/php/tests/CustomersModuleTest.php
<?php

declare(strict_types=1);

namespace Ecollect\Tests;

use Ecollect\Config;
use Ecollect\Exceptions\CustomerException;
use Ecollect\Exceptions\ValidationException;
use Ecollect\Modules\CustomersModule;
use Ecollect\Modules\SessionModule;
use Ecollect\Utils\HttpClient;
use PHPUnit\Framework\MockObject\MockObject;
use PHPUnit\Framework\TestCase;

class CustomersModuleTest extends TestCase
{
    private function makeConfig(): Config
    {
        return new Config([
            'api_key'     => 'test-key',
            'ety_code'    => 123,
            'environment' => 'test',
        ]);
    }

    /** @return SessionModule&MockObject */
    private function mockSession(string $token = 'session-token'): SessionModule
    {
        $mock = $this->createMock(SessionModule::class);
        $mock->method('getActive')->willReturn($token);
        return $mock;
    }

    private function customerInfo(): array
    {
        return [
            'email'               => 'user@example.com',
            'document_number'     => '12345678',
            'document_type'       => 'CC',
            'full_name'           => 'Juan Perez',
            'mobile_country_code' => '57',
            'mobile_number'       => '3001234567',
        ];
    }

    private function successResponse(string $customerId): array
    {
        return [
            'ReturnCode'        => 'SUCCESS',
            'CustomerInfoArray' => [
                ['AttributeCode' => 100, 'AttributeDesc' => 'CustomerId', 'AttributeValue' => $customerId],
            ],
        ];
    }

    public function testGetOrCreateCustomerIdNewCustomer(): void
    {
        $http = $this->createMock(HttpClient::class);
        $http->method('post')->willReturn($this->successResponse('cust-001'));

        $module = new CustomersModule($this->makeConfig(), $http, $this->mockSession());
        $result = $module->getOrCreateCustomerId($this->customerInfo());

        $this->assertSame('cust-001', $result['customerId']);
        $this->assertSame('user@example.com', $result['email']);
        $this->assertSame('Juan Perez', $result['fullName']);
    }

    public function testGetOrCreateCustomerIdExistingCustomer(): void
    {
        $http = $this->createMock(HttpClient::class);
        $http->method('post')->willReturn($this->successResponse('cust-existing-999'));

        $module = new CustomersModule($this->makeConfig(), $http, $this->mockSession());
        $result = $module->getOrCreateCustomerId($this->customerInfo());

        $this->assertSame('cust-existing-999', $result['customerId']);
    }

    public function testGetOrCreateRetrySucceedsAfterExpiredSession(): void
    {
        $http = $this->createMock(HttpClient::class);
        $http->expects($this->exactly(2))
            ->method('post')
            ->willReturnOnConsecutiveCalls(
                ['ReturnCode' => 'FAIL_APIEXPIREDSESSION'],
                $this->successResponse('cust-retry-001')
            );

        $session = $this->createMock(SessionModule::class);
        $session->method('getActive')->willReturn('session-token');

        $module = new CustomersModule($this->makeConfig(), $http, $session);
        $result = $module->getOrCreateCustomerId($this->customerInfo());

        $this->assertSame('cust-retry-001', $result['customerId']);
    }

    public function testGetOrCreateThrowsOnExpiredSessionRetryFails(): void
    {
        $http = $this->createMock(HttpClient::class);
        $http->expects($this->exactly(2))
            ->method('post')
            ->willReturnOnConsecutiveCalls(
                ['ReturnCode' => 'FAIL_APIEXPIREDSESSION'],
                ['ReturnCode' => 'FAIL_SYSTEM']
            );

        $session = $this->createMock(SessionModule::class);
        $session->method('getActive')->willReturn('session-token');

        $module = new CustomersModule($this->makeConfig(), $http, $session);

        $this->expectException(\Ecollect\Exceptions\NetworkRetryableException::class);
        $module->getOrCreateCustomerId($this->customerInfo());
    }

    public function testUpdateCustomerInfoSuccess(): void
    {
        $http = $this->createMock(HttpClient::class);
        $http->method('post')->willReturn([
            'ReturnCode'        => 'SUCCESS',
            'CustomerInfoArray' => [
                ['AttributeCode' => 100, 'AttributeDesc' => 'CustomerId', 'AttributeValue' => 'cust-001'],
            ],
        ]);

        $module = new CustomersModule($this->makeConfig(), $http, $this->mockSession());
        $result = $module->updateCustomerInfo('cust-001', [
            'email'     => 'new@example.com',
            'full_name' => 'Juan Updated',
        ]);

        $this->assertSame('cust-001', $result['customerId']);
        $this->assertSame('new@example.com', $result['email']);
    }

    public function testUpdateCustomerInfoThrowsOnNotFound(): void
    {
        $http = $this->createMock(HttpClient::class);
        $http->method('post')->willReturn(['ReturnCode' => 'FAIL_CARDHOLDERID']);

        $module = new CustomersModule($this->makeConfig(), $http, $this->mockSession());

        $this->expectException(ValidationException::class);
        $module->updateCustomerInfo('bad-id', ['email' => 'x@y.com']);
    }

    public function testGetOrCreateThrowsMissingCustomerIdInResponse(): void
    {
        $http = $this->createMock(HttpClient::class);
        $http->method('post')->willReturn([
            'ReturnCode'        => 'SUCCESS',
            'CustomerInfoArray' => [],
        ]);

        $module = new CustomersModule($this->makeConfig(), $http, $this->mockSession());

        $this->expectException(CustomerException::class);
        $module->getOrCreateCustomerId($this->customerInfo());
    }
}
